scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -164,123 +195,217 @@
+    ; We should have the cross product of
+    ;      :cook-expected-state/starting :cook-expected-state/running :cook-expected-state/completed :cook-expected-state/killed :missing
+    ; and
+    ;      :pod/waiting :pod/running :pod/succeeded :pod/failed :pod/unknown :missing
+    ;
+    ; for a total of 30 states.
+    ;
+    ; The only terminal expected states are :cook-expected-state/completed and :missing
+    ; The only terminal pod states are :pod/succeeded :pod/failed and :missing. We also treat :pod/unknown as a terminal state.
+
+    ; If you ignore the reloading on startup, the initial state is set at (:cook-expected-state/starting, :missing) when we first add a pod to launch.
+    ; The final state is (:missing, :missing)
+    ;
+    ; We use :cook-expected-state/killed to represent a user-chosen kill. If we're doing a state-machine-induced kill (because something
+    ; went wrong) it should occur by deleting the pod, so we go, e.g.,  (:running,:waiting) (an illegal state) to (:running,:missing)
+    ; to (:completed, :missing), to (:missing,missing) to deleted. We put a flag on when we delete so that we can indicate
+    ; the provenance (e.g., induced because of a weird state)
+    ;
+    ; Invariants:
+    ;   pod-has-just-completed/pod-was-killed/pod-has-started: These are callbacks invoked when kubernetes has moved
+    ;       to a new state. They handle writeback to datomic only. pod-was-killed is called in all of the weird
+    ;       kubernetes states that are not expected to occur, while pod-has-just-completed is invoked in all of the
+    ;       normal exit states (including exit-with-failure). I.e., these MUST only be invoked if cook expected state is not
+    ;       in a terminal state. These functions return new cook-expected-state-dict's so should be invoked last in almost
+    ;       all cases. (However, in a few cases, (delete-pod ? ?) can be invoked afterwards. It deletes the key from
+    ;       the cook expected state completely. Should only be invoked if we're nil.)
+    ;
+    ;   weird can be called at any time to indicate we're in a weird state and extra logging is warranted.
+    ;
+    ;   We always end with one of  pod-has-just-completed/pod-was-killed/pod-has-started or delete-pod.
+    ;
+    ;   We always update datomic first (with pod-has-* pod-was-*), etc, then we update kubernetes so we can handle restarts.
+    ;
+    ;   (delete-pod ? ?) is only valid when the kubernetes pod is in a terminal-state.
+    ;
+    ;   Note that we have release semantics. We could fail to process any operation. This means that, where relevant, we
+    ;     first write to datomic, then change into the next step. We thus show some intermediate states in the machine
+    ;     that seem like they could be skipped.
+    ;
+    ;   In the (:completed,*) states, we delete from datomic, transitioning into (:completed,:missing) and thence to (:missing, :missing)
+    ;     When we're in a terminal pod state.
```

Comment:
Is this part of the sentence on the previous line?

Reply:
Yes.